### PR TITLE
Incremental update

### PR DESCRIPTION
**Pull request by ZacLN**
@davidanthoff I've added a rudimentary partial update which _seems_  to work without problems. I couldn't think of any complicated corner cases beyond multiple cursor pasting to trick it up but you may want to have a test yourself before merging.

### CODE REVIEW

**Comment by ZacLN** [comment]
Incidentally we lose a bit of speed converting from `SubString` to `Vector{UInt8}`  and back again so we should probably shift to holding documents simply as raw vectors. The immutability of `String`s causes all sorts of trouble

File: scripts/languageserver/provider_misc.jl [hunk 1]
```diff
@@ -75,10 +75,36 @@ end
...
+        end
     try
-        documents[r.params.textDocument.uri] = split(r.params.contentChanges[1].text, r"\r\n?|\n")
         return
     catch err
         return Response{didChange,Exception}("2.0",r.id,err)
     end
 end
+
+function (c::TextDocumentContentChangeEvent)(vtdi::VersionedTextDocumentIdentifier)
```
Comment (davidanthoff) [inline scripts/languageserver/provider_misc.jl]: Could we make this a normal function? This just seems to complicate things...

File: scripts/languageserver/provider_misc.jl [hunk 2]
```diff
@@ -75,10 +75,36 @@ end
...
+    d = join(documents[vtdi.uri],'\n').data
```
Comment (davidanthoff) [inline scripts/languageserver/provider_misc.jl]: I think this is not very efficient. Why put everything together into one large string, if we can just modify those lines that are actually changed? I think the idea would be that most of these change requests probably only affect only one line, so those change requests can be handled very efficiently because you only need to touch one element in the array that holds the strings for a document.

Comment (ZacLN) [inline scripts/languageserver/provider_misc.jl]: I think I was working under the misapprehension that split resulted in a vector of `SubString`s, I'll have another look

File: scripts/languageserver/provider_misc.jl (hunk 1 above)
Comment (ZacLN) [inline scripts/languageserver/provider_misc.jl]: yep, sure. My whole repo has gone wonky so I'm going to close this and start afresh

File: scripts/languageserver/provider_misc.jl (hunk 2 above)
Comment (davidanthoff) [inline scripts/languageserver/provider_misc.jl]: Ah, you are right! Well, but what we could do is simply start out with an array of ``SubString``s, and then as we receive change notifications, we could replace those ``SubString``s with real ``String``s for those lines that get modified?

File: scripts/languageserver/provider_misc.jl (hunk 1 above)
Comment (davidanthoff) [inline scripts/languageserver/provider_misc.jl]: Sorry, that was probably partly my mistake by changing so much...

File: scripts/languageserver/provider_misc.jl (hunk 2 above)
Comment (ZacLN) [inline scripts/languageserver/provider_misc.jl]: Actually, looking at it the type specification on the document Dict seems to force a conversion when we open a file

Comment (ZacLN) [inline scripts/languageserver/provider_misc.jl]: I don't know much about String handling tbh, I've always had the impression it's all quite slow in Julia?. I'm going to do this properly (tediously) and set up and compare 4 approaches (full sync, Vector{UInt},, Vector{String} and Vector{SubString})

Comment (davidanthoff) [inline scripts/languageserver/provider_misc.jl]: ``Vector{UInt}`` would probably be tricky, then we would have to do the whole UTF-8 handling ourselves... Do we know whether this part is actually a real bottleneck? I would be kind of surprised. Right now even the full sync performs pretty well on my system, so I would think if we were to move to a ``Vector{AbstractString}`` thing, it could only get better.

Comment (ZacLN) [inline scripts/languageserver/provider_misc.jl]: Ok I'll just do that and look into it later
